// @subql/cli/lib/commands/validate.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const common_1 = require("@subql/common");
const validator_1 = require("@subql/validator");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
class Validate extends core_1.Command {
    //TODO, currently validation only work for complete project, ipfs deployment is not supported
    async run() {
        var _a;
        const { flags } = await this.parse(Validate);
        const location = (_a = flags.location) !== null && _a !== void 0 ? _a : process.cwd();
        const v = await validator_1.Validator.create(location, { ipfs: flags.ipfs }, flags['network-family']);
        const ipfsMatch = location.match(common_1.IPFS_REGEX);
        if (ipfsMatch) {
            v.addRule(...validator_1.deploymentRules);
        }
        else {
            v.addRule(...validator_1.commonRules);
        }
        const reports = await v.getValidateReports();
        const passed = reports.filter((r) => r.valid).length;
        const skipped = reports.filter((r) => r.skipped).length;
        const failed = reports.length - passed - skipped;
        if (!flags.silent) {
            for (const r of reports) {
                if (r.valid) {
                    this.log(`${chalk_1.default.bgGreen.bold(' PASS ')} ${r.name}`);
                }
                else if (r.skipped) {
                    this.log(`${chalk_1.default.yellow.bold(' SKIP ')} ${r.name}`);
                }
                else {
                    this.log(`${chalk_1.default.bgRed.bold(' FAIL ')} ${r.name}`);
                    this.log(`       ${chalk_1.default.redBright(r.description)}`);
                }
            }
            this.log('');
            this.log(`Result: ${passed} passed, ${failed} failed, ${skipped} skipped`);
            this.log('');
        }
        if (failed > 0) {
            this.exit(1);
        }
    }
}
exports.default = Validate;
Validate.description = 'Check a folder or github repo is a validate subquery project';
Validate.flags = {
    location: core_1.Flags.string({ char: 'l', description: 'local folder, github repo url or IPFS cid' }),
    ipfs: core_1.Flags.string({
        description: 'IPFS gateway endpoint, used for validating projects on IPFS',
        default: common_1.IPFS_NODE_ENDPOINT,
    }),
    silent: core_1.Flags.boolean(),
    'network-family': core_1.Flags.enum({ options: Object.values(common_1.NETWORK_FAMILY) }),
};
