// @subql/common-ethereum/dist/project/versioned/ProjectManifestVersioned.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumProjectManifestVersioned = exports.manifestIsV1_0_0 = exports.manifestIsV0_3_0 = void 0;
const class_transformer_1 = require("class-transformer");
const v0_3_0_1 = require("./v0_3_0");
const v1_0_0_1 = require("./v1_0_0");
const Ethereum_SUPPORTED_VERSIONS = {
    '0.3.0': v0_3_0_1.ProjectManifestV0_3_0Impl,
    '1.0.0': v1_0_0_1.ProjectManifestV1_0_0Impl,
};
function manifestIsV0_3_0(manifest) {
    return manifest.specVersion === '0.3.0';
}
exports.manifestIsV0_3_0 = manifestIsV0_3_0;
function manifestIsV1_0_0(manifest) {
    return manifest.specVersion === '1.0.0';
}
exports.manifestIsV1_0_0 = manifestIsV1_0_0;
class EthereumProjectManifestVersioned {
    constructor(projectManifest) {
        const klass = Ethereum_SUPPORTED_VERSIONS[projectManifest.specVersion];
        if (!klass) {
            throw new Error('specVersion not supported for project manifest file');
        }
        this._impl = (0, class_transformer_1.plainToClass)(klass, projectManifest);
    }
    get asImpl() {
        return this._impl;
    }
    get isV0_3_0() {
        return this.specVersion === '0.3.0';
    }
    get asV0_3_0() {
        return this._impl;
    }
    get isV1_0_0() {
        return this.specVersion === '1.0.0';
    }
    get asV1_0_0() {
        return this._impl;
    }
    toDeployment() {
        return this._impl.toDeployment();
    }
    validate() {
        return this._impl.validate();
    }
    get dataSources() {
        return this._impl.dataSources;
    }
    get schema() {
        return this._impl.schema.file;
    }
    get specVersion() {
        return this._impl.specVersion;
    }
    get description() {
        return this._impl.description;
    }
    get repository() {
        return this._impl.repository;
    }
}
exports.EthereumProjectManifestVersioned = EthereumProjectManifestVersioned;
