// @subql/common-near/dist/project/utils.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRuntimeDs = exports.isCustomDs = exports.isReceiptHandlerProcessor = exports.isActionHandlerProcessor = exports.isTransactionHandlerProcessor = exports.isBlockHandlerProcessor = void 0;
const types_near_1 = require("@subql/types-near");
function isBlockHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_near_1.NearHandlerKind.Block;
}
exports.isBlockHandlerProcessor = isBlockHandlerProcessor;
function isTransactionHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_near_1.NearHandlerKind.Transaction;
}
exports.isTransactionHandlerProcessor = isTransactionHandlerProcessor;
function isActionHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_near_1.NearHandlerKind.Action;
}
exports.isActionHandlerProcessor = isActionHandlerProcessor;
function isReceiptHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_near_1.NearHandlerKind.Receipt;
}
exports.isReceiptHandlerProcessor = isReceiptHandlerProcessor;
function isCustomDs(ds) {
    return ds.kind !== types_near_1.NearDatasourceKind.Runtime && !!ds.processor;
}
exports.isCustomDs = isCustomDs;
function isRuntimeDs(ds) {
    return ds.kind === types_near_1.NearDatasourceKind.Runtime;
}
exports.isRuntimeDs = isRuntimeDs;
