// @subql/cli/lib/commands/project/index.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const inquirer = tslib_1.__importStar(require("inquirer"));
const create_project_1 = tslib_1.__importDefault(require("./create-project"));
const delete_project_1 = tslib_1.__importDefault(require("./delete-project"));
class Project extends core_1.Command {
    async run() {
        const { flags } = await this.parse(Project);
        const option = flags.options;
        let userOptions;
        if (!option) {
            const response = await inquirer.prompt([
                {
                    name: 'projectOption',
                    message: 'Select an project option',
                    type: 'list',
                    choices: [{ name: 'create' }, { name: 'delete' }],
                },
            ]);
            userOptions = response.projectOption;
        }
        else {
            userOptions = option;
        }
        this.log(`Selected project option: ${userOptions}`);
        try {
            const handler = Project.optionMapping[userOptions];
            // removes arguments -> deployment and everything before it from the process.argv
            const stripped_argv = process.argv
                .filter((v, idx) => idx > process.argv.indexOf('project') && !v.includes('--options'))
                .reduce((acc, val) => acc.concat(val.split('=')), []);
            await handler.run(stripped_argv);
        }
        catch (e) {
            this.log(`Failed to execute command: ${userOptions} error: ${e}`);
        }
    }
}
exports.default = Project;
Project.description = 'Create/Delete project';
Project.flags = Object.assign(Object.assign({ options: core_1.Flags.string({
        options: ['create', 'delete'],
    }) }, delete_project_1.default.flags), create_project_1.default.flags);
Project.optionMapping = {
    create: create_project_1.default,
    delete: delete_project_1.default,
};
