// @subql/common-flare/dist/project/load.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadChainTypesFromJs = exports.parseChainTypes = exports.loadEthereumProjectManifest = exports.loadChainTypes = exports.parseEthereumProjectManifest = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const common_1 = require("@subql/common");
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const vm2_1 = require("vm2");
const models_1 = require("./models");
const versioned_1 = require("./versioned");
function parseEthereumProjectManifest(raw) {
    const projectManifest = new versioned_1.EthereumProjectManifestVersioned(raw);
    projectManifest.validate();
    return projectManifest;
}
exports.parseEthereumProjectManifest = parseEthereumProjectManifest;
function loadChainTypes(file, projectRoot) {
    const { ext } = path_1.default.parse(file);
    const filePath = path_1.default.resolve(projectRoot, file);
    if (fs_1.default.existsSync(filePath)) {
        if (ext === '.js' || ext === '.cjs') {
            //load can be self contained js file, or js depend on node_module which will require project root
            return loadChainTypesFromJs(filePath, projectRoot);
        }
        else if (ext === '.yaml' || ext === '.yml' || ext === '.json') {
            return (0, common_1.loadFromJsonOrYaml)(filePath);
        }
        else {
            throw new Error(`Extension ${ext} not supported`);
        }
    }
    else {
        throw new Error(`Load from file ${file} not exist`);
    }
}
exports.loadChainTypes = loadChainTypes;
function loadEthereumProjectManifest(file) {
    let manifestPath = file;
    if (fs_1.default.existsSync(file) && fs_1.default.lstatSync(file).isDirectory()) {
        const yamlFilePath = path_1.default.join(file, 'project.yaml');
        const jsonFilePath = path_1.default.join(file, 'project.json');
        if (fs_1.default.existsSync(yamlFilePath)) {
            manifestPath = yamlFilePath;
        }
        else if (fs_1.default.existsSync(jsonFilePath)) {
            manifestPath = jsonFilePath;
        }
        else {
            throw new Error(`Could not find project manifest under dir ${file}`);
        }
    }
    const doc = (0, common_1.loadFromJsonOrYaml)(manifestPath);
    const projectManifest = new versioned_1.EthereumProjectManifestVersioned(doc);
    projectManifest.validate();
    return projectManifest;
}
exports.loadEthereumProjectManifest = loadEthereumProjectManifest;
function parseChainTypes(raw) {
    const chainTypes = (0, class_transformer_1.plainToClass)(models_1.ChainTypes, raw);
    if (!!chainTypes.types ||
        !!chainTypes.typesChain ||
        !!chainTypes.typesBundle ||
        !!chainTypes.typesAlias ||
        !!chainTypes.typesSpec) {
        const errors = (0, class_validator_1.validateSync)(chainTypes, { whitelist: true, forbidNonWhitelisted: true });
        if (errors === null || errors === void 0 ? void 0 : errors.length) {
            // TODO: print error details
            const errorMsgs = errors.map((e) => e.toString()).join('\n');
            throw new Error(`failed to parse chain types.\n${errorMsgs}`);
        }
        return chainTypes;
    }
    else {
        throw new Error(`chainTypes is not valid`);
    }
}
exports.parseChainTypes = parseChainTypes;
function loadChainTypesFromJs(filePath, requireRoot) {
    const { base, ext } = path_1.default.parse(filePath);
    const root = requireRoot !== null && requireRoot !== void 0 ? requireRoot : path_1.default.dirname(filePath);
    const vm = new vm2_1.NodeVM({
        console: 'redirect',
        wasm: false,
        sandbox: {},
        require: {
            context: 'sandbox',
            external: true,
            builtin: ['path'],
            root: root,
            resolve: (moduleName) => {
                return require.resolve(moduleName, { paths: [root] });
            },
        },
        wrapper: 'commonjs',
        sourceExtensions: ['js', 'cjs'],
    });
    let rawContent;
    try {
        const script = new vm2_1.VMScript(`module.exports = require('${filePath}').default;`, path_1.default.join(root, 'sandbox')).compile();
        rawContent = vm.run(script);
    }
    catch (err) {
        throw new Error(`\n NodeVM error: ${err}`);
    }
    if (rawContent === undefined) {
        throw new Error(`There was no default export found from required ${base} file`);
    }
    return rawContent;
}
exports.loadChainTypesFromJs = loadChainTypesFromJs;
