// @subql/common-flare/dist/project/models.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomDataSourceBase = exports.FileReferenceImpl = exports.RuntimeDataSourceBase = exports.EthereumProcessorOptions = exports.CustomMapping = exports.EthereumMapping = exports.CustomHandler = exports.EventHandler = exports.CallHandler = exports.BlockHandler = exports.forbidNonWhitelisted = exports.TransactionFilter = exports.ChainTypes = exports.LogFilter = exports.BlockFilter = void 0;
const types_flare_1 = require("@subql/types-flare");
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const types_1 = require("./types");
class BlockFilter {
}
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInt)(),
    __metadata("design:type", Number)
], BlockFilter.prototype, "modulo", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], BlockFilter.prototype, "timestamp", void 0);
exports.BlockFilter = BlockFilter;
class LogFilter {
}
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsArray)(),
    __metadata("design:type", Array)
], LogFilter.prototype, "topics", void 0);
exports.LogFilter = LogFilter;
class ChainTypes {
}
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", Object)
], ChainTypes.prototype, "types", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", Object)
], ChainTypes.prototype, "typesAlias", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", Object)
], ChainTypes.prototype, "typesBundle", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", Object)
], ChainTypes.prototype, "typesChain", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", Object)
], ChainTypes.prototype, "typesSpec", void 0);
exports.ChainTypes = ChainTypes;
class TransactionFilter {
}
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], TransactionFilter.prototype, "from", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], TransactionFilter.prototype, "to", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], TransactionFilter.prototype, "function", void 0);
exports.TransactionFilter = TransactionFilter;
function forbidNonWhitelisted(keys, validationOptions) {
    return function (object, propertyName) {
        (0, class_validator_1.registerDecorator)({
            name: 'forbidNonWhitelisted',
            target: object.constructor,
            propertyName: propertyName,
            constraints: [],
            options: validationOptions,
            validator: {
                validate(value, args) {
                    const isValid = !Object.keys(value).some((key) => !(key in keys));
                    if (!isValid) {
                        throw new Error(`Invalid keys present in value: ${JSON.stringify(value)}. Whitelisted keys: ${JSON.stringify(Object.keys(keys))}`);
                    }
                    return isValid;
                },
            },
        });
    };
}
exports.forbidNonWhitelisted = forbidNonWhitelisted;
class BlockHandler {
}
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_transformer_1.Type)(() => BlockFilter),
    __metadata("design:type", BlockFilter)
], BlockHandler.prototype, "filter", void 0);
__decorate([
    (0, class_validator_1.IsEnum)(types_1.SubqlEthereumHandlerKind, { groups: [types_1.SubqlEthereumHandlerKind.FlareBlock, types_1.SubqlEthereumHandlerKind.EthBlock] }),
    __metadata("design:type", String)
], BlockHandler.prototype, "kind", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], BlockHandler.prototype, "handler", void 0);
exports.BlockHandler = BlockHandler;
class CallHandler {
}
__decorate([
    forbidNonWhitelisted({ from: '', to: '', function: '' }),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => TransactionFilter),
    __metadata("design:type", Object)
], CallHandler.prototype, "filter", void 0);
__decorate([
    (0, class_validator_1.IsEnum)(types_1.SubqlEthereumHandlerKind, { groups: [types_1.SubqlEthereumHandlerKind.FlareCall, types_1.SubqlEthereumHandlerKind.EthCall] }),
    __metadata("design:type", String)
], CallHandler.prototype, "kind", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CallHandler.prototype, "handler", void 0);
exports.CallHandler = CallHandler;
class EventHandler {
}
__decorate([
    forbidNonWhitelisted({ topics: '' }),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => LogFilter),
    __metadata("design:type", Object)
], EventHandler.prototype, "filter", void 0);
__decorate([
    (0, class_validator_1.IsEnum)(types_1.SubqlEthereumHandlerKind, { groups: [types_1.SubqlEthereumHandlerKind.FlareEvent, types_1.SubqlEthereumHandlerKind.EthEvent] }),
    __metadata("design:type", String)
], EventHandler.prototype, "kind", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], EventHandler.prototype, "handler", void 0);
exports.EventHandler = EventHandler;
class CustomHandler {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomHandler.prototype, "kind", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomHandler.prototype, "handler", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", Object)
], CustomHandler.prototype, "filter", void 0);
exports.CustomHandler = CustomHandler;
class EthereumMapping {
}
__decorate([
    (0, class_transformer_1.Transform)((params) => {
        const handlers = params.value;
        return handlers.map((handler) => {
            switch (handler.kind) {
                case types_1.SubqlEthereumHandlerKind.FlareEvent:
                    return (0, class_transformer_1.plainToClass)(EventHandler, handler);
                case types_1.SubqlEthereumHandlerKind.FlareCall:
                    return (0, class_transformer_1.plainToClass)(CallHandler, handler);
                case types_1.SubqlEthereumHandlerKind.FlareBlock:
                    return (0, class_transformer_1.plainToClass)(BlockHandler, handler);
                case types_1.SubqlEthereumHandlerKind.EthEvent:
                    return (0, class_transformer_1.plainToClass)(EventHandler, handler);
                case types_1.SubqlEthereumHandlerKind.EthCall:
                    return (0, class_transformer_1.plainToClass)(CallHandler, handler);
                case types_1.SubqlEthereumHandlerKind.EthBlock:
                    return (0, class_transformer_1.plainToClass)(BlockHandler, handler);
                default:
                    throw new Error(`handler ${handler.kind} not supported`);
            }
        });
    }),
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", Array)
], EthereumMapping.prototype, "handlers", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], EthereumMapping.prototype, "file", void 0);
exports.EthereumMapping = EthereumMapping;
class CustomMapping {
}
__decorate([
    (0, class_validator_1.IsArray)(),
    (0, class_transformer_1.Type)(() => CustomHandler),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", Array)
], CustomMapping.prototype, "handlers", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomMapping.prototype, "file", void 0);
exports.CustomMapping = CustomMapping;
class EthereumProcessorOptions {
}
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], EthereumProcessorOptions.prototype, "abi", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsEthereumAddress)(),
    __metadata("design:type", String)
], EthereumProcessorOptions.prototype, "address", void 0);
exports.EthereumProcessorOptions = EthereumProcessorOptions;
class RuntimeDataSourceBase {
}
__decorate([
    (0, class_validator_1.IsEnum)(types_1.SubqlEthereumDatasourceKind, {
        groups: [types_1.SubqlEthereumDatasourceKind.FlareRuntime, types_1.SubqlEthereumDatasourceKind.EthRuntime],
    }),
    __metadata("design:type", String)
], RuntimeDataSourceBase.prototype, "kind", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => EthereumMapping),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", Object)
], RuntimeDataSourceBase.prototype, "mapping", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInt)(),
    __metadata("design:type", Number)
], RuntimeDataSourceBase.prototype, "startBlock", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", Map)
], RuntimeDataSourceBase.prototype, "assets", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => EthereumProcessorOptions),
    __metadata("design:type", EthereumProcessorOptions)
], RuntimeDataSourceBase.prototype, "options", void 0);
exports.RuntimeDataSourceBase = RuntimeDataSourceBase;
class FileReferenceImpl {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], FileReferenceImpl.prototype, "file", void 0);
exports.FileReferenceImpl = FileReferenceImpl;
class CustomDataSourceBase {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomDataSourceBase.prototype, "kind", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => CustomMapping),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", Object)
], CustomDataSourceBase.prototype, "mapping", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInt)(),
    __metadata("design:type", Number)
], CustomDataSourceBase.prototype, "startBlock", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => FileReferenceImpl),
    (0, class_validator_1.ValidateNested)({ each: true }),
    __metadata("design:type", Map)
], CustomDataSourceBase.prototype, "assets", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => FileReferenceImpl),
    (0, class_validator_1.IsObject)(),
    __metadata("design:type", Object)
], CustomDataSourceBase.prototype, "processor", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", EthereumProcessorOptions)
], CustomDataSourceBase.prototype, "options", void 0);
exports.CustomDataSourceBase = CustomDataSourceBase;
