// @subql/common-flare/dist/project/types.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubqlEthereumDatasourceKind = exports.SubqlEthereumHandlerKind = exports.EthereumDatasourceKind = exports.EthereumHandlerKind = void 0;
// All of these used to be redefined in this file, re-exporting for simplicity
var types_flare_1 = require("@subql/types-flare");
Object.defineProperty(exports, "EthereumHandlerKind", { enumerable: true, get: function () { return types_flare_1.EthereumHandlerKind; } });
Object.defineProperty(exports, "EthereumDatasourceKind", { enumerable: true, get: function () { return types_flare_1.EthereumDatasourceKind; } });
var SubqlEthereumHandlerKind;
(function (SubqlEthereumHandlerKind) {
    SubqlEthereumHandlerKind["FlareBlock"] = "flare/BlockHandler";
    SubqlEthereumHandlerKind["FlareCall"] = "flare/TransactionHandler";
    SubqlEthereumHandlerKind["FlareEvent"] = "flare/LogHandler";
    SubqlEthereumHandlerKind["EthBlock"] = "ethereum/BlockHandler";
    SubqlEthereumHandlerKind["EthCall"] = "ethereum/TransactionHandler";
    SubqlEthereumHandlerKind["EthEvent"] = "ethereum/LogHandler";
})(SubqlEthereumHandlerKind = exports.SubqlEthereumHandlerKind || (exports.SubqlEthereumHandlerKind = {}));
var SubqlEthereumDatasourceKind;
(function (SubqlEthereumDatasourceKind) {
    SubqlEthereumDatasourceKind["FlareRuntime"] = "flare/Runtime";
    SubqlEthereumDatasourceKind["EthRuntime"] = "ethereum/Runtime";
})(SubqlEthereumDatasourceKind = exports.SubqlEthereumDatasourceKind || (exports.SubqlEthereumDatasourceKind = {}));
