// @subql/cli/lib/commands/deployment/index.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const inquirer = tslib_1.__importStar(require("inquirer"));
const delete_1 = tslib_1.__importDefault(require("./delete"));
const deploy_1 = tslib_1.__importDefault(require("./deploy"));
const promote_1 = tslib_1.__importDefault(require("./promote"));
class Deployment extends core_1.Command {
    async run() {
        const { flags } = await this.parse(Deployment);
        const option = flags.options;
        let userOptions;
        if (!option) {
            const response = await inquirer.prompt([
                {
                    name: 'deploymentOptions',
                    message: 'Select an deployment option',
                    type: 'list',
                    choices: [{ name: 'deploy' }, { name: 'promote' }, { name: 'delete' }],
                },
            ]);
            userOptions = response.deploymentOptions;
        }
        else {
            userOptions = option;
        }
        this.log(`Selected deployment option: ${userOptions}`);
        try {
            const handler = Deployment.optionMapping[userOptions];
            // removes arguments -> deployment and everything before it from the process.argv
            const stripped_argv = process.argv
                .filter((v, idx) => idx > process.argv.indexOf('deployment') && !v.includes('--options'))
                .reduce((acc, val) => acc.concat(val.split('=')), []);
            await handler.run(stripped_argv);
        }
        catch (e) {
            this.log(`Failed to execute command: ${userOptions} error: ${e}`);
        }
    }
}
exports.default = Deployment;
Deployment.description = 'Deploy to hosted service';
Deployment.flags = Object.assign(Object.assign(Object.assign({ options: core_1.Flags.string({
        options: ['deploy', 'promote', 'delete'],
    }) }, deploy_1.default.flags), promote_1.default.flags), delete_1.default.flags);
Deployment.optionMapping = {
    deploy: deploy_1.default,
    promote: promote_1.default,
    delete: delete_1.default,
};
