// @subql/common-near/dist/project/versioned/v0_2_1/model.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectManifestV0_2_1Impl = exports.DeploymentV0_2_1 = exports.CustomDatasourceTemplateImpl = exports.RuntimeDatasourceTemplateImpl = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const v0_2_0_1 = require("../v0_2_0");
class RuntimeDatasourceTemplateImpl extends v0_2_0_1.NearRuntimeDataSourceV0_2_0Impl {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], RuntimeDatasourceTemplateImpl.prototype, "name", void 0);
exports.RuntimeDatasourceTemplateImpl = RuntimeDatasourceTemplateImpl;
class CustomDatasourceTemplateImpl extends v0_2_0_1.NearCustomDataSourceV0_2_0Impl {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomDatasourceTemplateImpl.prototype, "name", void 0);
exports.CustomDatasourceTemplateImpl = CustomDatasourceTemplateImpl;
class DeploymentV0_2_1 extends v0_2_0_1.DeploymentV0_2_0 {
}
__decorate([
    (0, class_validator_1.Equals)('0.2.1'),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], DeploymentV0_2_1.prototype, "specVersion", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CustomDatasourceTemplateImpl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: RuntimeDatasourceTemplateImpl, name: 'near/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], DeploymentV0_2_1.prototype, "templates", void 0);
exports.DeploymentV0_2_1 = DeploymentV0_2_1;
class ProjectManifestV0_2_1Impl extends v0_2_0_1.ProjectManifestV0_2_0Impl {
    get deployment() {
        if (!this._deployment) {
            this._deployment = (0, class_transformer_1.plainToClass)(DeploymentV0_2_1, this);
            (0, class_validator_1.validateSync)(this._deployment, { whitelist: true });
        }
        return this._deployment;
    }
}
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CustomDatasourceTemplateImpl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: RuntimeDatasourceTemplateImpl, name: 'near/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], ProjectManifestV0_2_1Impl.prototype, "templates", void 0);
exports.ProjectManifestV0_2_1Impl = ProjectManifestV0_2_1Impl;
