// @subql/common-algorand/dist/project/models.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomDataSourceBase = exports.FileReferenceImpl = exports.RuntimeDataSourceBase = exports.CustomMapping = exports.RuntimeMapping = exports.CustomHandler = exports.TransactionHandler = exports.BlockHandler = exports.TransactionFilter = exports.BlockFilter = void 0;
const types_algorand_1 = require("@subql/types-algorand");
const algosdk_1 = require("algosdk");
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
class BlockFilter {
}
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInt)(),
    __metadata("design:type", Number)
], BlockFilter.prototype, "modulo", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], BlockFilter.prototype, "timestamp", void 0);
exports.BlockFilter = BlockFilter;
class TransactionFilter {
}
__decorate([
    (0, class_validator_1.IsEnum)(algosdk_1.TransactionType),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", String)
], TransactionFilter.prototype, "txType", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateIf)((o) => {
        return o.txType === algosdk_1.TransactionType.pay || o.txType === algosdk_1.TransactionType.axfer || o.txType === algosdk_1.TransactionType.appl;
    }),
    __metadata("design:type", String)
], TransactionFilter.prototype, "sender", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateIf)((o) => {
        return o.txType === algosdk_1.TransactionType.pay || o.txType === algosdk_1.TransactionType.axfer;
    }),
    __metadata("design:type", String)
], TransactionFilter.prototype, "receiver", void 0);
__decorate([
    (0, class_validator_1.IsBoolean)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateIf)((o) => o.txType === algosdk_1.TransactionType.keyreg),
    __metadata("design:type", Boolean)
], TransactionFilter.prototype, "nonParticipant", void 0);
__decorate([
    (0, class_validator_1.IsNumber)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateIf)((o) => {
        return o.txType === algosdk_1.TransactionType.acfg || o.txType === algosdk_1.TransactionType.axfer || o.txType === algosdk_1.TransactionType.afrz;
    }),
    __metadata("design:type", Number)
], TransactionFilter.prototype, "assetId", void 0);
__decorate([
    (0, class_validator_1.IsBoolean)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateIf)((o) => o.txType === algosdk_1.TransactionType.afrz),
    __metadata("design:type", Boolean)
], TransactionFilter.prototype, "newFreezeStatus", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateIf)((o) => o.txType === algosdk_1.TransactionType.afrz),
    __metadata("design:type", String)
], TransactionFilter.prototype, "address", void 0);
__decorate([
    (0, class_validator_1.IsNumber)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateIf)((o) => o.txType === algosdk_1.TransactionType.appl),
    __metadata("design:type", Number)
], TransactionFilter.prototype, "applicationId", void 0);
exports.TransactionFilter = TransactionFilter;
class BlockHandler {
}
__decorate([
    (0, class_validator_1.IsEnum)(types_algorand_1.AlgorandHandlerKind, { groups: [types_algorand_1.AlgorandHandlerKind.Block] }),
    __metadata("design:type", String)
], BlockHandler.prototype, "kind", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], BlockHandler.prototype, "handler", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => BlockFilter),
    __metadata("design:type", Object)
], BlockHandler.prototype, "filter", void 0);
exports.BlockHandler = BlockHandler;
class TransactionHandler {
}
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => TransactionFilter),
    __metadata("design:type", Object)
], TransactionHandler.prototype, "filter", void 0);
__decorate([
    (0, class_validator_1.IsEnum)(types_algorand_1.AlgorandHandlerKind, { groups: [types_algorand_1.AlgorandHandlerKind.Block] }),
    __metadata("design:type", String)
], TransactionHandler.prototype, "kind", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], TransactionHandler.prototype, "handler", void 0);
exports.TransactionHandler = TransactionHandler;
class CustomHandler {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomHandler.prototype, "kind", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomHandler.prototype, "handler", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", Object)
], CustomHandler.prototype, "filter", void 0);
exports.CustomHandler = CustomHandler;
class RuntimeMapping {
}
__decorate([
    (0, class_transformer_1.Transform)((params) => {
        const handlers = params.value;
        return handlers.map((handler) => {
            switch (handler.kind) {
                case types_algorand_1.AlgorandHandlerKind.Block:
                    return (0, class_transformer_1.plainToClass)(BlockHandler, handler);
                case types_algorand_1.AlgorandHandlerKind.Transaction:
                    return (0, class_transformer_1.plainToClass)(TransactionHandler, handler);
                default:
                    throw new Error(`handler ${handler.kind} not supported`);
            }
        });
    }),
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", Array)
], RuntimeMapping.prototype, "handlers", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], RuntimeMapping.prototype, "file", void 0);
exports.RuntimeMapping = RuntimeMapping;
class CustomMapping {
}
__decorate([
    (0, class_validator_1.IsArray)(),
    (0, class_transformer_1.Type)(() => CustomHandler),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", Array)
], CustomMapping.prototype, "handlers", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomMapping.prototype, "file", void 0);
exports.CustomMapping = CustomMapping;
class RuntimeDataSourceBase {
}
__decorate([
    (0, class_validator_1.IsEnum)(types_algorand_1.AlgorandDataSourceKind, { groups: [types_algorand_1.AlgorandDataSourceKind.Runtime] }),
    __metadata("design:type", String)
], RuntimeDataSourceBase.prototype, "kind", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => RuntimeMapping),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", RuntimeMapping)
], RuntimeDataSourceBase.prototype, "mapping", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInt)(),
    __metadata("design:type", Number)
], RuntimeDataSourceBase.prototype, "startBlock", void 0);
exports.RuntimeDataSourceBase = RuntimeDataSourceBase;
class FileReferenceImpl {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], FileReferenceImpl.prototype, "file", void 0);
exports.FileReferenceImpl = FileReferenceImpl;
class CustomDataSourceBase {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomDataSourceBase.prototype, "kind", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => CustomMapping),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", Object)
], CustomDataSourceBase.prototype, "mapping", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInt)(),
    __metadata("design:type", Number)
], CustomDataSourceBase.prototype, "startBlock", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => FileReferenceImpl),
    (0, class_validator_1.ValidateNested)({ each: true }),
    __metadata("design:type", Map)
], CustomDataSourceBase.prototype, "assets", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => FileReferenceImpl),
    (0, class_validator_1.IsObject)(),
    __metadata("design:type", Object)
], CustomDataSourceBase.prototype, "processor", void 0);
exports.CustomDataSourceBase = CustomDataSourceBase;
