// @subql/common-algorand/dist/project/validation/is-string-or-object.validation.js
"use strict";
// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.IsStringOrObject = void 0;
const class_validator_1 = require("class-validator");
function IsStringOrObject(validationOptions) {
    return function (object, propertyName) {
        (0, class_validator_1.registerDecorator)({
            name: 'IsStringOrObject',
            target: object.constructor,
            propertyName: propertyName,
            constraints: [],
            options: validationOptions,
            validator: {
                validate(value) {
                    return typeof value === 'string' || (typeof value === 'object' && Object.keys(value).length > 0);
                },
                defaultMessage() {
                    throw new Error(`${propertyName} must by string or object!`);
                },
            },
        });
    };
}
exports.IsStringOrObject = IsStringOrObject;
