// @subql/common-substrate/dist/project/utils.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSubstrateTemplates = exports.isRuntimeDs = exports.isCustomDs = exports.isCallHandlerProcessor = exports.isEventHandlerProcessor = exports.isBlockHandlerProcessor = void 0;
const types_1 = require("@subql/types");
const semver_1 = require("semver");
function isBlockHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_1.SubstrateHandlerKind.Block;
}
exports.isBlockHandlerProcessor = isBlockHandlerProcessor;
function isEventHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_1.SubstrateHandlerKind.Event;
}
exports.isEventHandlerProcessor = isEventHandlerProcessor;
function isCallHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_1.SubstrateHandlerKind.Call;
}
exports.isCallHandlerProcessor = isCallHandlerProcessor;
function isCustomDs(ds) {
    return ds.kind !== types_1.SubstrateDatasourceKind.Runtime && !!ds.processor;
}
exports.isCustomDs = isCustomDs;
function isRuntimeDs(ds) {
    return ds.kind === types_1.SubstrateDatasourceKind.Runtime;
}
exports.isRuntimeDs = isRuntimeDs;
function isSubstrateTemplates(templatesData, specVersion) {
    return (isRuntimeDs(templatesData[0]) || isCustomDs(templatesData[0])) && (0, semver_1.gte)(specVersion, '0.2.1');
}
exports.isSubstrateTemplates = isSubstrateTemplates;
