// @polkadot/util-crypto/cjs/mnemonic/bip39.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateMnemonic = exports.generateMnemonic = exports.entropyToMnemonic = exports.mnemonicToEntropy = exports.mnemonicToSeedSync = void 0;
const tslib_1 = require("tslib");
const util_1 = require("@polkadot/util");
const index_js_1 = require("../pbkdf2/index.js");
const index_js_2 = require("../random/index.js");
const index_js_3 = require("../sha/index.js");
const bip39_en_js_1 = tslib_1.__importDefault(require("./bip39-en.js"));
const INVALID_MNEMONIC = 'Invalid mnemonic';
const INVALID_ENTROPY = 'Invalid entropy';
const INVALID_CHECKSUM = 'Invalid mnemonic checksum';
function normalize(str) {
    return (str || '').normalize('NFKD');
}
function binaryToByte(bin) {
    return parseInt(bin, 2);
}
function bytesToBinary(bytes) {
    return bytes.map((x) => x.toString(2).padStart(8, '0')).join('');
}
function deriveChecksumBits(entropyBuffer) {
    return bytesToBinary(Array.from((0, index_js_3.sha256AsU8a)(entropyBuffer))).slice(0, (entropyBuffer.length * 8) / 32);
}
function mnemonicToSeedSync(mnemonic, password) {
    return (0, index_js_1.pbkdf2Encode)((0, util_1.stringToU8a)(normalize(mnemonic)), (0, util_1.stringToU8a)(`mnemonic${normalize(password)}`)).password;
}
exports.mnemonicToSeedSync = mnemonicToSeedSync;
function mnemonicToEntropy(mnemonic) {
    const words = normalize(mnemonic).split(' ');
    if (words.length % 3 !== 0) {
        throw new Error(INVALID_MNEMONIC);
    }
    // convert word indices to 11 bit binary strings
    const bits = words
        .map((word) => {
        const index = bip39_en_js_1.default.indexOf(word);
        if (index === -1) {
            throw new Error(INVALID_MNEMONIC);
        }
        return index.toString(2).padStart(11, '0');
    })
        .join('');
    // split the binary string into ENT/CS
    const dividerIndex = Math.floor(bits.length / 33) * 32;
    const entropyBits = bits.slice(0, dividerIndex);
    const checksumBits = bits.slice(dividerIndex);
    // calculate the checksum and compare
    const matched = entropyBits.match(/(.{1,8})/g);
    const entropyBytes = matched && matched.map(binaryToByte);
    if (!entropyBytes || (entropyBytes.length % 4 !== 0) || (entropyBytes.length < 16) || (entropyBytes.length > 32)) {
        throw new Error(INVALID_ENTROPY);
    }
    const entropy = (0, util_1.u8aToU8a)(entropyBytes);
    if (deriveChecksumBits(entropy) !== checksumBits) {
        throw new Error(INVALID_CHECKSUM);
    }
    return entropy;
}
exports.mnemonicToEntropy = mnemonicToEntropy;
function entropyToMnemonic(entropy) {
    // 128 <= ENT <= 256
    if ((entropy.length % 4 !== 0) || (entropy.length < 16) || (entropy.length > 32)) {
        throw new Error(INVALID_ENTROPY);
    }
    const matched = `${bytesToBinary(Array.from(entropy))}${deriveChecksumBits(entropy)}`.match(/(.{1,11})/g);
    const mapped = matched && matched.map((binary) => bip39_en_js_1.default[binaryToByte(binary)]);
    if (!mapped || (mapped.length < 12)) {
        throw new Error('Unable to map entropy to mnemonic');
    }
    return mapped.join(' ');
}
exports.entropyToMnemonic = entropyToMnemonic;
function generateMnemonic(numWords) {
    return entropyToMnemonic((0, index_js_2.randomAsU8a)((numWords / 3) * 4));
}
exports.generateMnemonic = generateMnemonic;
function validateMnemonic(mnemonic) {
    try {
        mnemonicToEntropy(mnemonic);
    }
    catch {
        return false;
    }
    return true;
}
exports.validateMnemonic = validateMnemonic;
