// @subql/common/dist/constants.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.runnerMapping = exports.NETWORK_FAMILY = exports.MMR_AWAIT_TIME = exports.DEFAULT_LEAF = exports.DEFAULT_WORD_SIZE = exports.RUNNER_REGEX = exports.IPFS_REGEX = exports.IPFS_CLUSTER_ENDPOINT = exports.IPFS_NODE_ENDPOINT = exports.IPFS_WRITE = exports.IPFS_READ = void 0;
//IPFS
exports.IPFS_READ = 'https://unauthipfs.subquery.network';
exports.IPFS_WRITE = 'https://authipfs.subquery.network';
exports.IPFS_NODE_ENDPOINT = `${exports.IPFS_READ}/ipfs/api/v0`;
exports.IPFS_CLUSTER_ENDPOINT = `${exports.IPFS_WRITE}/cluster/add`;
exports.IPFS_REGEX = /^ipfs:\/\//i;
// MANIFEST
exports.RUNNER_REGEX = /(\^?)(\d|x|\*)+\.(\d|x|\*)+\.(\d|x|\*)+/;
//MMR
exports.DEFAULT_WORD_SIZE = 32;
exports.DEFAULT_LEAF = Buffer.from('0000000000000000000000000000000000000000000000000000000000000001', 'hex');
exports.MMR_AWAIT_TIME = 2;
// NETWORK
var NETWORK_FAMILY;
(function (NETWORK_FAMILY) {
    NETWORK_FAMILY["substrate"] = "Substrate";
    NETWORK_FAMILY["avalanche"] = "Avalanche";
    NETWORK_FAMILY["terra"] = "Terra";
    NETWORK_FAMILY["cosmos"] = "Cosmos";
    NETWORK_FAMILY["algorand"] = "Algorand";
})(NETWORK_FAMILY = exports.NETWORK_FAMILY || (exports.NETWORK_FAMILY = {}));
exports.runnerMapping = {
    '@subql/node': NETWORK_FAMILY.substrate,
    '@subql/node-substrate': NETWORK_FAMILY.substrate,
    '@subql/node-avalanche': NETWORK_FAMILY.avalanche,
    '@subql/node-terra': NETWORK_FAMILY.terra,
    '@subql/node-cosmos': NETWORK_FAMILY.cosmos,
    '@subql/node-algorand': NETWORK_FAMILY.algorand,
};
