// @subql/cli/lib/controller/project-controller.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteProject = exports.createProject = exports.suffixFormat = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const utils_1 = require("../utils");
const suffixFormat = (value) => {
    return value
        .replace(/(^\s*)|(\s*$)/g, '')
        .replace(/\s+/g, '-')
        .toLowerCase();
};
exports.suffixFormat = suffixFormat;
async function createProject(organization, subtitle, logoUrl, project_name, authToken, gitRepository, description, apiVersion, dedicateDB, url) {
    try {
        const result = (await (0, axios_1.default)({
            headers: {
                Authorization: `Bearer ${authToken}`,
            },
            method: 'post',
            url: 'subqueries',
            baseURL: url,
            data: {
                apiVersion: `v${apiVersion}`,
                description: description,
                gitRepository: gitRepository,
                key: `${organization}/${(0, exports.suffixFormat)(project_name)}`,
                logoUrl: logoUrl,
                name: project_name,
                subtitle: subtitle,
                dedicateDBKey: dedicateDB,
            },
        })).data;
        return result;
    }
    catch (e) {
        (0, utils_1.errorHandle)(e, 'Failed to create project:');
    }
}
exports.createProject = createProject;
async function deleteProject(authToken, organization, project_name, url) {
    const key = `${organization}/${project_name}`;
    try {
        await (0, axios_1.default)({
            headers: {
                Authorization: `Bearer ${authToken}`,
            },
            method: 'delete',
            url: `subqueries/${key}`,
            baseURL: url,
        });
        return `${key}`;
    }
    catch (e) {
        (0, utils_1.errorHandle)(e, 'Failed to delete project:');
    }
}
exports.deleteProject = deleteProject;
