// @subql/cli/lib/controller/project-controller.spec.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../constants");
const utils_1 = require("../utils");
const project_controller_1 = require("./project-controller");
const projectSpec = {
    org: process.env.SUBQL_ORG_TEST,
    projectName: 'mocked_project',
    repository: 'https://github.com/bz888/test-deployment-2',
    ipfs: 'QmaVh8DGzuRCJZ5zYEDxXQsXYqP9HihjjeuxNNteSDq8xX',
    subtitle: '',
    description: '',
    logoURl: '',
    apiVersion: '2',
};
// Replace/Update your access token when test locally
const testAuth = process.env.SUBQL_ACCESS_TOKEN_TEST;
const testIf = (condition, ...args) => (condition ? it(...args) : it.skip(...args));
jest.setTimeout(120000);
describe('CLI create project and delete project', () => {
    testIf(!!testAuth, 'Create project and delete', async () => {
        const { apiVersion, description, logoURl, org, projectName, repository, subtitle } = projectSpec;
        const create_project = await (0, project_controller_1.createProject)(org, subtitle, logoURl, projectName, testAuth, repository, description, apiVersion, undefined, constants_1.ROOT_API_URL_DEV);
        await (0, utils_1.delay)(10);
        const delete_project = await (0, project_controller_1.deleteProject)(testAuth, org, projectName, constants_1.ROOT_API_URL_DEV);
        expect(create_project.key).toMatch(`${process.env.SUBQL_ORG_TEST}/mocked_project`);
        expect(delete_project).toMatch(`${process.env.SUBQL_ORG_TEST}/mocked_project`);
    });
});
