// @subql/common-cosmos/dist/project/utils.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCosmosTemplates = exports.isEventHandlerProcessor = exports.isMessageHandlerProcessor = exports.isTransactionHandlerProcessor = exports.isBlockHandlerProcessor = exports.isRuntimeCosmosDs = exports.isCustomCosmosDs = void 0;
const types_cosmos_1 = require("@subql/types-cosmos");
const semver_1 = require("semver");
function isCustomCosmosDs(ds) {
    return ds.kind !== types_cosmos_1.SubqlCosmosDatasourceKind.Runtime && !!ds.processor;
}
exports.isCustomCosmosDs = isCustomCosmosDs;
function isRuntimeCosmosDs(ds) {
    return ds.kind === types_cosmos_1.SubqlCosmosDatasourceKind.Runtime;
}
exports.isRuntimeCosmosDs = isRuntimeCosmosDs;
function isBlockHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_cosmos_1.SubqlCosmosHandlerKind.Block;
}
exports.isBlockHandlerProcessor = isBlockHandlerProcessor;
function isTransactionHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_cosmos_1.SubqlCosmosHandlerKind.Transaction;
}
exports.isTransactionHandlerProcessor = isTransactionHandlerProcessor;
function isMessageHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_cosmos_1.SubqlCosmosHandlerKind.Message;
}
exports.isMessageHandlerProcessor = isMessageHandlerProcessor;
function isEventHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_cosmos_1.SubqlCosmosHandlerKind.Event;
}
exports.isEventHandlerProcessor = isEventHandlerProcessor;
function isCosmosTemplates(templatesData, specVersion) {
    return (isRuntimeCosmosDs(templatesData[0]) || isCustomCosmosDs(templatesData[0])) && (0, semver_1.gte)(specVersion, '0.2.1');
}
exports.isCosmosTemplates = isCosmosTemplates;
