// @subql/common-terra/dist/project/utils.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTerraTemplates = exports.isEventHandlerProcessor = exports.isMessageHandlerProcessor = exports.isTransactionHandlerProcessor = exports.isBlockHandlerProcessor = exports.isRuntimeTerraDs = exports.isCustomTerraDs = void 0;
const types_terra_1 = require("@subql/types-terra");
const semver_1 = require("semver");
function isCustomTerraDs(ds) {
    return ds.kind !== types_terra_1.SubqlTerraDatasourceKind.Runtime && !!ds.processor;
}
exports.isCustomTerraDs = isCustomTerraDs;
function isRuntimeTerraDs(ds) {
    return ds.kind === types_terra_1.SubqlTerraDatasourceKind.Runtime;
}
exports.isRuntimeTerraDs = isRuntimeTerraDs;
function isBlockHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_terra_1.SubqlTerraHandlerKind.Block;
}
exports.isBlockHandlerProcessor = isBlockHandlerProcessor;
function isTransactionHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_terra_1.SubqlTerraHandlerKind.Transaction;
}
exports.isTransactionHandlerProcessor = isTransactionHandlerProcessor;
function isMessageHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_terra_1.SubqlTerraHandlerKind.Message;
}
exports.isMessageHandlerProcessor = isMessageHandlerProcessor;
function isEventHandlerProcessor(hp) {
    return hp.baseHandlerKind === types_terra_1.SubqlTerraHandlerKind.Event;
}
exports.isEventHandlerProcessor = isEventHandlerProcessor;
function isTerraTemplates(templatesData, specVersion) {
    return (isRuntimeTerraDs(templatesData[0]) || isCustomTerraDs(templatesData[0])) && (0, semver_1.gte)(specVersion, '1.0.0');
}
exports.isTerraTemplates = isTerraTemplates;
