// @subql/common-cosmos/dist/project/versioned/v0_3_0/model.js
"use strict";
// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectManifestV0_3_0Impl = exports.DeploymentV0_3_0 = exports.CosmosCustomDataSourceV0_3_0Impl = exports.CosmosRuntimeDataSourceV0_3_0Impl = exports.CosmosProjectMappingV0_3_0 = exports.CosmosProjectNetworkV0_3_0 = exports.CosmosProjectNetworkDeploymentV0_3_0 = exports.CustomModuleImpl = exports.FileTypeV0_3_0 = void 0;
const common_1 = require("@subql/common");
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const models_1 = require("../../models");
class FileTypeV0_3_0 {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], FileTypeV0_3_0.prototype, "file", void 0);
exports.FileTypeV0_3_0 = FileTypeV0_3_0;
class CustomModuleImpl {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CustomModuleImpl.prototype, "file", void 0);
__decorate([
    (0, class_validator_1.IsArray)(),
    __metadata("design:type", Array)
], CustomModuleImpl.prototype, "messages", void 0);
exports.CustomModuleImpl = CustomModuleImpl;
class CosmosProjectNetworkDeploymentV0_3_0 {
}
__decorate([
    (0, class_validator_1.IsString)(),
    (0, class_validator_1.IsNotEmpty)(),
    (0, class_transformer_1.Transform)(({ value }) => value.trim()),
    __metadata("design:type", String)
], CosmosProjectNetworkDeploymentV0_3_0.prototype, "chainId", void 0);
__decorate([
    (0, class_transformer_1.Type)(() => models_1.CosmosCustomModuleImpl),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.ValidateNested)({ each: true }),
    __metadata("design:type", Map)
], CosmosProjectNetworkDeploymentV0_3_0.prototype, "chainTypes", void 0);
__decorate([
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsArray)(),
    __metadata("design:type", Array)
], CosmosProjectNetworkDeploymentV0_3_0.prototype, "bypassBlocks", void 0);
exports.CosmosProjectNetworkDeploymentV0_3_0 = CosmosProjectNetworkDeploymentV0_3_0;
class CosmosProjectNetworkV0_3_0 extends CosmosProjectNetworkDeploymentV0_3_0 {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CosmosProjectNetworkV0_3_0.prototype, "endpoint", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", String)
], CosmosProjectNetworkV0_3_0.prototype, "dictionary", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", String)
], CosmosProjectNetworkV0_3_0.prototype, "genesisHash", void 0);
exports.CosmosProjectNetworkV0_3_0 = CosmosProjectNetworkV0_3_0;
class CosmosProjectMappingV0_3_0 extends models_1.CosmosMapping {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], CosmosProjectMappingV0_3_0.prototype, "file", void 0);
exports.CosmosProjectMappingV0_3_0 = CosmosProjectMappingV0_3_0;
class CosmosRuntimeDataSourceV0_3_0Impl extends models_1.CosmosRuntimeDataSourceBase {
}
__decorate([
    (0, class_transformer_1.Type)(() => CosmosProjectMappingV0_3_0),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", Object)
], CosmosRuntimeDataSourceV0_3_0Impl.prototype, "mapping", void 0);
exports.CosmosRuntimeDataSourceV0_3_0Impl = CosmosRuntimeDataSourceV0_3_0Impl;
class CosmosCustomDataSourceV0_3_0Impl extends models_1.CosmosCustomDataSourceBase {
}
exports.CosmosCustomDataSourceV0_3_0Impl = CosmosCustomDataSourceV0_3_0Impl;
class DeploymentV0_3_0 {
}
__decorate([
    (0, class_validator_1.Equals)('0.3.0'),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], DeploymentV0_3_0.prototype, "specVersion", void 0);
__decorate([
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => FileTypeV0_3_0),
    __metadata("design:type", FileTypeV0_3_0)
], DeploymentV0_3_0.prototype, "schema", void 0);
__decorate([
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CosmosCustomDataSourceV0_3_0Impl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: CosmosRuntimeDataSourceV0_3_0Impl, name: 'cosmos/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], DeploymentV0_3_0.prototype, "dataSources", void 0);
__decorate([
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CosmosProjectNetworkDeploymentV0_3_0),
    __metadata("design:type", CosmosProjectNetworkDeploymentV0_3_0)
], DeploymentV0_3_0.prototype, "network", void 0);
exports.DeploymentV0_3_0 = DeploymentV0_3_0;
class ProjectManifestV0_3_0Impl extends common_1.ProjectManifestBaseImpl {
    get deployment() {
        if (!this._deployment) {
            this._deployment = (0, class_transformer_1.plainToClass)(DeploymentV0_3_0, this);
            (0, class_validator_1.validateSync)(this._deployment, { whitelist: true });
        }
        return this._deployment;
    }
}
__decorate([
    (0, class_validator_1.Equals)('0.3.0'),
    __metadata("design:type", String)
], ProjectManifestV0_3_0Impl.prototype, "specVersion", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProjectManifestV0_3_0Impl.prototype, "name", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProjectManifestV0_3_0Impl.prototype, "version", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CosmosProjectNetworkV0_3_0),
    __metadata("design:type", CosmosProjectNetworkV0_3_0)
], ProjectManifestV0_3_0Impl.prototype, "network", void 0);
__decorate([
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => FileTypeV0_3_0),
    __metadata("design:type", FileTypeV0_3_0)
], ProjectManifestV0_3_0Impl.prototype, "schema", void 0);
__decorate([
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CosmosCustomDataSourceV0_3_0Impl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: CosmosRuntimeDataSourceV0_3_0Impl, name: 'cosmos/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], ProjectManifestV0_3_0Impl.prototype, "dataSources", void 0);
exports.ProjectManifestV0_3_0Impl = ProjectManifestV0_3_0Impl;
